biorxiv/pca_association_experiment/biorxiv_pca_category_bootstrap.py

```diff
@@ -0,0 +1,230 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# # Determine Word to PCA Associations
+
+# This notebook is designed to run PCA over the document embeddings and calculate category-pca associations with each principal component.
+
+# In[1]:
+
+
+from pathlib import Path
+
+import numpy as np
+import pandas as pd
+import plotnine as p9
+from scipy.spatial.distance import cdist
+from sklearn.decomposition import PCA
+from tqdm import tqdm_notebook
+
+
+# In[2]:
+
+
+journal_map_df = pd.read_csv(
+    Path("..")/
+    Path("exploratory_data_analysis")/
+    Path("output/biorxiv_article_metadata.tsv"), 
+    sep="\t"
+)
+journal_map_df.head()
+
+
+# # PCA the Documents
+
+# Run PCA over the documents. Generates 50 principal components, but can generate more or less.
+
+# In[3]:
+
+
+n_components = 50
+random_state = 100
+
+
+# In[4]:
+
+
+biorxiv_articles_df = pd.read_csv(
+    Path("..")/
+    Path("word_vector_experiment")/
+    Path("output/word2vec_output")/
+    Path("biorxiv_all_articles_300.tsv.xz"), 
+    sep="\t"
+)
+biorxiv_articles_df.head()
+
+
+# In[5]:
+
+
+reducer = PCA(
+    n_components = n_components,
+    random_state = random_state
+)
+
+reducer.fit(
+    biorxiv_articles_df[[f"feat_{idx}" for idx in range(300)]]
+    .values
+)
+
+
+# # Bootstrap 95% Confidence Intervals Cosine sim
+
+# In[6]:
+
+
+document_categories_df = (
+    journal_map_df[["document", "category"]]
+    .merge(biorxiv_articles_df, on="document")
+)
+document_categories_df.head()
+
+
+# In[7]:
+
+
+boostraped_iterations = 10000
+pc_1 = []
+pc_2 = []
+
+for iteration in tqdm_notebook(range(boostraped_iterations)):
+    sampled_df = (
+        document_categories_df
+        .groupby("category")
+        .apply(lambda x: x.sample(1, random_state=iteration))
+        .reset_index(drop=True)
+        .sort_values("category")
+    )
+    
+    document_distance = (
+        1 - cdist(
+            sampled_df
+            .drop(["document", "category"], axis=1)
+            .values, 
+            reducer.components_[0:2], 
+            'cosine'
+        )
+    )
+    
+    pc_1.append(document_distance[:, 0])
+    pc_2.append(document_distance[:, 1])
+
+pc_1 = np.stack(pc_1)
+pc_2 = np.stack(pc_2)
+
+
+# In[8]:
+
+
+mean_similarity_df = (
+    pd.DataFrame(
+        (
+            1 - cdist(
+                document_categories_df
+                .drop(["document", "category"], axis=1)
+                .values, 
+                reducer.components_[0:2], 
+                'cosine'
+            )
+        ),
+        columns=["pca1_cossim", "pca2_cossim"]
+    )
+    .assign(
+        category = document_categories_df.category.tolist(),
+        document = document_categories_df.document.tolist()
+    )
+)
+mean_similarity_df.head()
+
+
+# In[9]:
+
+
+category_sim_df = (
```

Comment:
Here is where you're combining everything, so you are taking the mean similarity of docs within a category and getting the CI for the similarity of docs in the same category (but i thought the iterations above were comparing between categories not within so not sure how this is a CI)

Reply:
> (but i thought the iterations above were comparing between categories not within so not sure how this is a CI)

No the similarity scores are between embeddings and principal components. After all the iterations have been generated, I'm taking the 25th and 97.5th percentiles to be my 95% confidence intervals for each category. This is a brute force way of estimating CI when I don't know the underlying distribution (called Bootstrapping).

Reply:
This is much clearer now that I understand the earlier bits of code. Thanks